src/Util/Store.hs

```diff
@@ -0,0 +1,96 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE MultiParamTypeClasses  #-}
+{-# LANGUAGE RoleAnnotations    #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE TypeFamilies       #-}
+{-# LANGUAGE TypeSynonymInstances #-}
+{-# LANGUAGE RankNTypes         #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE FlexibleContexts #-}
+
+-- | wrapper over IntMap for our purposes
+
+module Util.Store
+  ( lookup
+  , singleton
+  , insert
+  , toList
+  , fromList
+  , Store
+  , unionWith
+  , mapKeys
+  , mapMaybeWithKey
+  )
+where
+
+import Prelude hiding (lookup)
+import Control.Lens
+import Data.Data (Data)
+import Data.IntMap.Strict (IntMap)
+import qualified Data.IntMap.Strict as IM
+import Control.Arrow (first)
+
+import Util.Key
+import Phase
+
+newtype Store p v = Store { unStore :: IntMap v}
+  deriving newtype (Eq, Ord, Show, Semigroup, Monoid, Functor, Foldable)
+  deriving stock   Data
+type role Store representational _
```

Comment:
Why `representational`? I would have either have expected `phantom`, because IntMap's invariants only rely on Int's instances and not `p`'s, or `nominal`, to allow coercing a `Store K1 (Sum Int)` into a  `Store K1 Int` but prevent accidentally coercing a `Store K1 Int` into a `Store K2 Int`.

Reply:
(and similarly for `Util.Set`)

Reply:
I think nominal is better.